File: frontend/types/report.ts
/**
 * Work Reports (Daily & Weekly) Types
 * Aligned with backend app/schemas/report.py
 */

export type DailyReportStatus = 'DRAFT' | 'SUBMITTED' | 'REVIEWED' | 'RETURNED' | 'FLAGGED';
export type WeeklyReportStatus = 'DRAFT' | 'SUBMITTED' | 'REVIEWED' | 'RETURNED';

export interface DailyReportTaskCreate {
  task_id: string;
  progress_notes?: string;
}

export interface DailyReportTaskResponse {
  task_id: string;
  task_title?: string;
  task_status?: string;
  progress_notes?: string;
}

export interface DailyReportHistoryResponse {
  id: string;
  report_id: string;
  actor_id?: string;
  actor_username?: string;
  action: string;
  comments?: string;
  created_at: string;
}

export interface DailyReportResponse {
  id: string;
  user_id: string;
  author_id?: string;
  user_role?: string;
  username?: string;
  user_full_name?: string;
  vertical_id: string;
  vertical_name?: string;
  report_date: string;
  work_summary: string;
  tasks_completed?: string;
  tasks?: DailyReportTaskResponse[];
  blockers?: string;
  issues?: string;
  next_actions?: string;
  evidence_links?: string;
  status: DailyReportStatus;
  reviewer_id?: string;
  reviewer_username?: string;
  reviewed_by_id?: string;
  reviewed_by_username?: string;
  review_comments?: string;
  history?: DailyReportHistoryResponse[];
  submitted_at?: string;
  reviewed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface DailyReportListResponse {
  total: number;
  items: DailyReportResponse[];
}

export interface DailyReportCreate {
  vertical_id?: string | null;
  report_date?: string | null;
  work_summary: string;
  tasks?: DailyReportTaskCreate[];
  task_ids?: string[];
  assigned_task_id?: string | null;
  tasks_completed?: string | null;
  blockers?: string | null;
  issues?: string | null;
  next_actions?: string | null;
  evidence_links?: string | null;
  submit_now?: boolean;
}

export interface DailyReportUpdate {
  work_summary?: string;
  tasks?: DailyReportTaskCreate[];
  task_ids?: string[];
  tasks_completed?: string | null;
  blockers?: string | null;
  issues?: string | null;
  next_actions?: string | null;
  evidence_links?: string | null;
  submit_now?: boolean;
}

export interface DailyReportReviewRequest {
  status: DailyReportStatus;
  review_comments?: string | null;
}

export interface WeeklyTaskSummary {
  id: string;
  title: string;
  status: string;
  priority: string;
  assigned_to_name?: string;
  deadline?: string;
}

export interface WeeklyIssueSummary {
  id: string;
  title: string;
  status: string;
  sensitivity: string;
}

export interface WeeklyRollupResponse {
  start_date: string;
  end_date: string;
  vertical_id?: string;
  vertical_name?: string;
  user_id?: string;
  user_name?: string;
  daily_reports_count: number;
  daily_reports_submitted: DailyReportResponse[];
  completed_tasks_count: number;
  completed_tasks: WeeklyTaskSummary[];
  incomplete_tasks_count: number;
  incomplete_tasks: WeeklyTaskSummary[];
  blockers_count?: number;
  blockers?: string[];
  major_issues_count?: number;
  major_issues?: WeeklyIssueSummary[];
  achievements?: string[];
  existing_weekly_report?: WeeklyReportResponse | null;
}

export interface WeeklyReportResponse {
  id: string;
  user_id: string;
  author_id?: string;
  user_role?: string;
  username?: string;
  user_full_name?: string;
  vertical_id: string;
  vertical_name?: string;
  week_start_date: string;
  week_end_date: string;
  days_reported_count?: number;
  days_reported?: any[];
  summary: string;
  completed_work?: string;
  outstanding_work?: string;
  tasks_worked_on?: DailyReportTaskResponse[];
  daily_reports?: DailyReportResponse[];
  blockers?: string;
  issues?: string;
  priorities_next_week?: string;
  supervisor_comments?: string;
  reviewer_id?: string;
  reviewer_username?: string;
  reviewed_by_id?: string;
  reviewed_by_username?: string;
  status: WeeklyReportStatus;
  submitted_at?: string;
  reviewed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface WeeklyReportListResponse {
  total: number;
  items: WeeklyReportResponse[];
}

export interface WeeklyReportReviewRequest {
  status: WeeklyReportStatus;
  supervisor_comments?: string | null;
}

File: app/schemas/report.py
"""
Pydantic Schemas for Daily & Weekly Work Reports & Weekly Rollup
Paradox Sports OMS - Phase 3 Core Operational System & Phase 10J Review Hierarchy
"""

from datetime import date, datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field, field_validator
from app.models.report import DailyReportStatus, WeeklyReportStatus


class DailyReportTaskCreate(BaseModel):
    task_id: UUID
    progress_notes: Optional[str] = None

    @field_validator("progress_notes", mode="before")
    @classmethod
    def coerce_empty_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class DailyReportTaskResponse(BaseModel):
    task_id: UUID
    task_title: Optional[str] = None
    task_status: Optional[str] = None
    progress_notes: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class DailyReportHistoryResponse(BaseModel):
    id: UUID
    report_id: UUID
    actor_id: Optional[UUID] = None
    actor_username: Optional[str] = None
    action: str
    comments: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DailyReportCreate(BaseModel):
    vertical_id: Optional[UUID] = None
    report_date: Optional[date] = None
    work_summary: str = Field(..., min_length=5)
    tasks: Optional[List[DailyReportTaskCreate]] = Field(default_factory=list)
    task_ids: Optional[List[UUID]] = Field(default_factory=list)
    assigned_task_id: Optional[UUID] = None  # backward compatibility
    tasks_completed: Optional[str] = None
    blockers: Optional[str] = None
    issues: Optional[str] = None
    next_actions: Optional[str] = None
    evidence_links: Optional[str] = None
    submit_now: bool = True

    @field_validator(
        "vertical_id",
        "report_date",
        "assigned_task_id",
        "tasks_completed",
        "blockers",
        "issues",
        "next_actions",
        "evidence_links",
        mode="before",
    )
    @classmethod
    def coerce_empty_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class DailyReportUpdate(BaseModel):
    work_summary: Optional[str] = Field(None, min_length=5)
    tasks: Optional[List[DailyReportTaskCreate]] = None
    task_ids: Optional[List[UUID]] = None
    tasks_completed: Optional[str] = None
    blockers: Optional[str] = None
    issues: Optional[str] = None
    next_actions: Optional[str] = None
    evidence_links: Optional[str] = None
    submit_now: Optional[bool] = None

    @field_validator(
        "tasks_completed",
        "blockers",
        "issues",
        "next_actions",
        "evidence_links",
        mode="before",
    )
    @classmethod
    def coerce_empty_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class DailyReportReviewRequest(BaseModel):
    status: DailyReportStatus  # REVIEWED, RETURNED, FLAGGED
    review_comments: Optional[str] = None

    @field_validator("review_comments", mode="before")
    @classmethod
    def coerce_empty_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class DailyReportResponse(BaseModel):
    id: UUID
    user_id: UUID
    author_id: Optional[UUID] = None
    user_role: Optional[str] = None
    username: Optional[str] = None
    user_full_name: Optional[str] = None
    vertical_id: UUID
    vertical_name: Optional[str] = None
    report_date: date
    work_summary: str
    tasks_completed: Optional[str] = None
    tasks: List[DailyReportTaskResponse] = Field(default_factory=list)
    blockers: Optional[str] = None
    issues: Optional[str] = None
    next_actions: Optional[str] = None
    evidence_links: Optional[str] = None
    status: DailyReportStatus
    reviewer_id: Optional[UUID] = None
    reviewer_username: Optional[str] = None
    reviewed_by_id: Optional[UUID] = None
    reviewed_by_username: Optional[str] = None
    review_comments: Optional[str] = None
    history: List[DailyReportHistoryResponse] = Field(default_factory=list)
    submitted_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DailyReportListResponse(BaseModel):
    total: int
    items: List[DailyReportResponse]


# Dynamic Weekly Rollup Models
class WeeklyTaskSummary(BaseModel):
    id: UUID
    title: str
    status: str
    priority: str
    assigned_to_name: Optional[str] = None
    deadline: Optional[datetime] = None


class WeeklyIssueSummary(BaseModel):
    id: UUID
    title: str
    status: str
    sensitivity: str


class WeeklyRollupResponse(BaseModel):
    start_date: date
    end_date: date
    vertical_id: Optional[UUID] = None
    vertical_name: Optional[str] = None
    user_id: Optional[UUID] = None
    user_name: Optional[str] = None
    daily_reports_count: int
    daily_reports_submitted: List[DailyReportResponse] = Field(default_factory=list)
    completed_tasks_count: int
    completed_tasks: List[WeeklyTaskSummary] = Field(default_factory=list)
    incomplete_tasks_count: int
    incomplete_tasks: List[WeeklyTaskSummary] = Field(default_factory=list)
    blockers_count: int
    blockers: List[str] = Field(default_factory=list)
    major_issues_count: int
    major_issues: List[WeeklyIssueSummary] = Field(default_factory=list)
    achievements: List[str] = Field(default_factory=list)
    existing_weekly_report: Optional["WeeklyReportResponse"] = None


class WeeklyReportCreate(BaseModel):
    vertical_id: Optional[UUID] = None
    week_start_date: date
    week_end_date: date
    summary: str = Field(..., min_length=5)
    completed_work: Optional[str] = None
    outstanding_work: Optional[str] = None
    blockers: Optional[str] = None
    issues: Optional[str] = None
    priorities_next_week: Optional[str] = None
    submit_now: bool = True

    @field_validator(
        "vertical_id",
        "completed_work",
        "outstanding_work",
        "blockers",
        "issues",
        "priorities_next_week",
        mode="before",
    )
    @classmethod
    def coerce_empty_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class WeeklyReportReviewRequest(BaseModel):
    status: WeeklyReportStatus  # REVIEWED, RETURNED
    supervisor_comments: Optional[str] = None

    @field_validator("supervisor_comments", mode="before")
    @classmethod
    def coerce_empty_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class WeeklyReportResponse(BaseModel):
    id: UUID
    user_id: UUID
    author_id: Optional[UUID] = None
    user_role: Optional[str] = None
    username: Optional[str] = None
    user_full_name: Optional[str] = None
    vertical_id: UUID
    vertical_name: Optional[str] = None
    week_start_date: date
    week_end_date: date
    days_reported_count: int = 0
    days_reported: List[Dict[str, Any]] = Field(default_factory=list)
    summary: str
    completed_work: Optional[str] = None
    outstanding_work: Optional[str] = None
    tasks_worked_on: List[DailyReportTaskResponse] = Field(default_factory=list)
    daily_reports: List[DailyReportResponse] = Field(default_factory=list)
    blockers: Optional[str] = None
    issues: Optional[str] = None
    priorities_next_week: Optional[str] = None
    supervisor_comments: Optional[str] = None
    reviewer_id: Optional[UUID] = None
    reviewer_username: Optional[str] = None
    reviewed_by_id: Optional[UUID] = None
    reviewed_by_username: Optional[str] = None
    status: WeeklyReportStatus
    submitted_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WeeklyReportListResponse(BaseModel):
    total: int
    items: List[WeeklyReportResponse]
